jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -2579,6 +2581,189 @@ ecma_builtin_array_prototype_object_copy_within (const ecma_value_t args[], /**<
+    if (ECMA_IS_VALUE_ERROR (element))
+    {
+      return element;
+    }
+
+    if (ecma_is_value_found (element))
+    {
+      if (is_mapped_value_not_undefined)
+      {
+        ecma_value_t sourceVal = ecma_make_uint32_value (sourceIndex);
+        ecma_value_t args[] = {element,  sourceVal, ecma_make_object_value (source)};
+        ecma_value_t temp_element = ecma_op_function_call (ecma_get_object_from_value (mapped_value), thisArg, args, 3);
+
+        ecma_free_value (element);
+        ecma_free_value (sourceVal);
+
+        if (ECMA_IS_VALUE_ERROR (temp_element))
+        {
+          return temp_element;
+        }
+
+        element = temp_element;
+      }
+      bool shouldFlatten = false;
+      if (depth > 0)
+      {
+        ecma_value_t is_array = ecma_is_value_array (element);
+
+        if (ECMA_IS_VALUE_ERROR (is_array))
+        {
+          ecma_free_value (element);
+          return is_array;
+        }
+        shouldFlatten = ecma_is_value_true (is_array);
+      }
+
+      if (shouldFlatten)
+      {
+        ecma_object_t *element_obj = ecma_get_object_from_value (element);
+        uint32_t elementLen;
+        ecma_value_t len_value = ecma_op_object_get_length (element_obj, &elementLen);
+
+        if (ECMA_IS_VALUE_ERROR (len_value))
+        {
+          ecma_deref_object (element_obj);
+          return len_value;
+        }
+
+        targetIndex = ecma_builtin_array_flatten_into_array (target,
+                                                             element_obj,
+                                                             elementLen,
+                                                             targetIndex,
+                                                             depth-1,
+                                                             ECMA_VALUE_UNDEFINED,
+                                                             ECMA_VALUE_UNDEFINED);
+        ecma_deref_object (element_obj);
+      }
+      else
+      {
+        ecma_value_t element_temp = ecma_builtin_helper_def_prop_by_index (ecma_get_object_from_value (target),
+                                                                           targetIndex,
+                                                                           element,
+                                                                      ECMA_PROPERTY_CONFIGURABLE_ENUMERABLE_WRITABLE);
+
+        ecma_free_value (element);
+        if (ECMA_IS_VALUE_ERROR (element_temp))
+        {
+          return element_temp;
+        }
+
+        targetIndex++;
+      }
+    }
+    sourceIndex++;
+  }
+  return targetIndex;
+} /* ecma_builtin_array_flatten_into_array */
+
+/**
+ * The Array.prototype object's 'flat' routine
+ *
+ * See also:
+ *          ECMA-262 v10, 22.1.3.10
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_array_prototype_object_flat (const ecma_value_t args[], /**< arguments list */
```

Comment:
Based on the standard each method (flat/flatMap can return errors during the processing).